include/openthread/diag.h

```diff
@@ -104,6 +104,26 @@ otError otDiagProcessCmdLine(otInstance *aInstance, const char *aString, char *a
  */
 bool otDiagIsEnabled(otInstance *aInstance);
 
+/**
+ * Pointer to callback to output diag messages.
+ *
+ * @param[in]  aFormat     The format string.
+ * @param[in]  aArguments  The format string arguments.
+ * @param[out] aContext    A pointer to the user context.
+ *
+ */
+typedef void (*otDiagOutputCallback)(const char *aFormat, va_list aArguments, void *aContext);
+
+/**
+ * Sets the diag output callback.
+ *
+ * @param[in]  aInstance   The OpenThread instance structure.
+ * @param[in]  aCallback   A pointer to a function that is called on outputting diag messages.
+ * @param[in]  aContext    A pointer to the user context.
+ *
+ */
+void otDiagSetOutputCallback(otInstance *aInstance, otDiagOutputCallback aCallback, void *aContext);
```

Comment:
It is a good idea to remove the `aOutput ` and `aOutputMaxLen` from the `otDiagProcessCmdLine ` and `otDiagProcessCmd`.

There is a limitation to add the `otDiagOutputCallback` to the `otDiagProcessCmdLine` and `otDiagProcessCmd`: After the function `otDiagProcessCmdLine` or `otDiagProcessCmd` returns, the diag module no longer be able to send out any diag message. This may be a problem when developers want to see the real-time output result when the diag module is running in background. For example: developers may want to see the RSSI value of each received frame. Current implementation of diag module is unable to implement this feature. Thoughts?